gax/src/test/java/com/google/api/gax/rpc/ServerStreamTest.java

```diff
@@ -0,0 +1,201 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.SettableApiFuture;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Queues;
+import com.google.common.truth.Truth;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class ServerStreamTest {
+  private TestStreamController controller;
+  private ServerStream<Integer> stream;
+  private ExecutorService executor;
+
+  @Rule public ExpectedException expectedException = ExpectedException.none();
+
+  @Before
+  public void setUp() throws Exception {
+    controller = new TestStreamController();
+    stream = new ServerStream<>();
+
+    stream.observer().onStart(controller);
+    executor = Executors.newSingleThreadExecutor();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    executor.shutdownNow();
+  }
+
+  @Test
+  public void testEmptyStream() {
+    stream.observer().onStart(controller);
+    stream.observer().onComplete();
+
+    Truth.assertThat(Lists.newArrayList(stream)).isEmpty();
+  }
+
+  @Test
+  public void testMultipleItemStream() throws InterruptedException {
+    executor.submit(
+        new Callable<Void>() {
+          @Override
+          public Void call() throws InterruptedException {
+            for (int i = 0; i < 5; i++) {
+              int requestCount = controller.requests.poll(1, TimeUnit.SECONDS);
+
+              Truth.assertWithMessage("ServerStream should request one item at a time")
```

Comment:
I don't think this does what you want. If the test fails, `Truth` throws, but the exception will be caught by the thread pool and logged; it won't propagate to the test thread and won't fail the test. The test will fail "eventually" but probably because the stream or the test times out.

If you want this to fail the test, I think you need to send the message back to the test thread, probably by a blocking queue?

Reply:
fixed